src/dbus_api/blockdev.rs

```diff
@@ -0,0 +1,276 @@
+/// found by means of the getter method which takes a reference to a
+/// blockdev and obtains the property from the blockdev.
+fn get_blockdev_property<F>(i: &mut IterAppend,
+                            p: &PropInfo<MTFn<TData>, TData>,
+                            getter: F)
+                            -> Result<(), MethodErr>
+    where F: Fn(&BlockDev) -> Result<MessageItem, MethodErr>
+{
+    let dbus_context = p.tree.get_data();
+    let object_path = p.path.get_name();
+
+    let blockdev_path = p.tree
+        .get(object_path)
+        .expect("tree must contain implicit argument");
+
+    let blockdev_data =
+        blockdev_path
+            .get_data()
+            .as_ref()
+            .ok_or_else(|| MethodErr::failed(&format!("no data for object path {}", object_path)))?;
+
+    let pool_path = p.tree
+        .get(&blockdev_data.parent)
+        .ok_or_else(|| {
+                        MethodErr::failed(&format!("no path for parent object path {}",
+                                                   &blockdev_data.parent))
+                    })?;
+
+    let pool_uuid = pool_path
+        .get_data()
+        .as_ref()
+        .ok_or_else(|| MethodErr::failed(&format!("no data for object path {}", object_path)))?
+        .uuid;
+
+    let engine = dbus_context.engine.borrow();
+    let pool =
+        engine
+            .get_pool(&pool_uuid)
+            .ok_or_else(|| {
+                            MethodErr::failed(&format!("no pool corresponding to uuid {}",
+                                                       &pool_uuid))
+                        })?;
+    let blockdev_uuid = &blockdev_data.uuid;
+    let blockdev =
+        pool.get_blockdev(blockdev_uuid)
+            .ok_or_else(|| MethodErr::failed(&format!("noblockdev with uuid {}", &blockdev_uuid)))?;
+    i.append(getter(blockdev)?);
+    Ok(())
+}
+
+/// Get the devnode for an object path.
+fn get_blockdev_devnode(i: &mut IterAppend,
+                        p: &PropInfo<MTFn<TData>, TData>)
+                        -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(format!("{}", p.devnode().display()))))
+}
+
+fn get_blockdev_hardware_info(i: &mut IterAppend,
+                              p: &PropInfo<MTFn<TData>, TData>)
+                              -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(p.hardware_info().unwrap_or("").to_owned())))
+}
+
+fn get_blockdev_user_info(i: &mut IterAppend,
+                          p: &PropInfo<MTFn<TData>, TData>)
+                          -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(p.user_info().unwrap_or("").to_owned())))
+}
+
+fn get_blockdev_initialization_time(i: &mut IterAppend,
```

Comment:
Maybe initialization time should actually just be a number, and client should do any formatting.

Reply:
seconds since the epoch? Fine with me.